Refactor cmake

kdt3rd opened the pull request:
This refactors the cmake setup, modernizing it to a current flavor of
cmake and cleaning up the definitions. This also makes the top level
folder a "super project", meaning it is including what should be
distinct / standalone sub-projects with their own finds that should
work.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Refactor cmake

This refactors the cmake setup, modernizing it to a current flavor of
cmake and cleaning up the definitions. This also makes the top level
folder a "super project", meaning it is including what should be
distinct / standalone sub-projects with their own finds that should
work.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd commented:
fyi, this is not yet ready for merge, as it does not yet include PyIlmBase or OpenEXR_viewers, but if people could review it, and then with different cmake configurations, windows, macs, etc. could test, would be much appreciated.

kdt3rd commented:
They will need testing, but this should resolve the following issues:
#363 
#362 
#357 
#343 
#333 
#164 
#125 
#123

kdt3rd committed:
Add missing file, remove unused exrbuild tool cmake

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix new (WIP) cmake setup to work on OS/X

This includes a fix for the semaphore configure check as well as a
couple of compile warnings

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
First pass making cross compile work, cross compiling windows using mingw on linux

This currently works for building using static libraries, but not yet
tested with dlls.

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Switch testing control to use standard ctest setting option

kdt3rd committed:
First pass updating the documentation for cmake builds

cary-ilm commented:
LGTM so far. I say merge it, and we can fix it from there.

kdt3rd committed:
Properly include additional cmake files in "make dist" under autoconf

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Generate version file to ensure proper version check

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Enable custom install subfolder for headers

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Documentation pass

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Attempt to set rpath for more common scenarios when people are building custom versions

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Start to integrate python module using FindPython and FindBoost from modern cmake

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula reviewed (approved):
Looking good so far.

kdt3rd committed:
Fix issues with rpath, message cleanup, checkpoint better python layer

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

waebbl commented on PyIlmBase/PyIexTest/CMakeLists.txt:
```diff
@@ -1,5 +1,18 @@
+  )
+endif()
+
+if(TARGET Python2::Interpreter)
```
Could this be a typo? I.e. it should read
```
if(TARGET Python3::Interpreter)
```

kdt3rd replied:
it was, fixed in the latest set of finishing up the refactor. thanks for noticing!

kdt3rd committed:
Remove un-needed files now that cmake can provide correct values

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
PyIlmBase finished refactor, misc cleanup

- add extra dist to automake for make dist
- finish numpy lookup
- add sample vfx 15 toolchain file for doc purposes
- merge cxx standard, pay attention to global setting if set
- merge clang tidy option
- add default build type if not set

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove removed file, add CMakeLists.txt file

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula commented on IlmBase/config/IlmBaseSetup.cmake:
```diff
@@ -10,7 +10,12 @@ option(ILMBASE_ENABLE_LARGE_STACK "Enables code to take advantage of large stack
 # What C++ standard to compile for
 # VFX Platform 18 is c++14, so let's enable that by default
-set(ILMBASE_CXX_STANDARD "14" CACHE STRING "C++ standard to compile against")
+set(tmp 14)
```
nice, useful pattern

meshula commented on PyIlmBase/PyImathNumpy/CMakeLists.txt:
```diff
@@ -1,36 +1,47 @@
+if(TARGET Python2::Python AND
```
yikes, this is complex, I'm looking forward to pybind11

kdt3rd replied:
yes, boost, and boost python is a nightmare, and then numpy is annoying because it's not part of boost python until 1.63. Python is a bit complex handling 2 and 3 at the same time, but that's a useful thing for VFX Platform 2020...
